Sources/Workspace/Workspace.swift

```diff
@@ -2619,6 +2632,48 @@ extension Workspace {
             return path
         }
     }
+    
+    private func checkSourceControlIntegrity(package: PackageReference,
+                                             version: Version,
+                                             revision: Revision,
+                                             observabilityScope: ObservabilityScope) throws {
+        guard case .remoteSourceControl(let sourceControlURL) = package.kind, let fingerprintStorage = self.fingerprintStorage else {
+            return
+        }
+        
+        let packageIdentity = package.identity
+        do {
+            let fingerprint = try temp_await {
+                fingerprintStorage.get(
+                    package: packageIdentity,
+                    version: version,
+                    kind: .sourceControl,
+                    observabilityScope: observabilityScope,
+                    callbackQueue: .sharedConcurrent,
+                    callback: $0)
+            }
+            
+            // The revision (i.e., hash) must match that in fingerprint storage otherwise the integrity check fails
+            guard revision.identifier == fingerprint.value else {
+                throw StringError("Source control fingerprint \(revision.identifier) for \(package) version=\(version) does not match previously recorded value \(fingerprint.value)")
+            }
+        } catch PackageFingerprintStorageError.notFound {
+            // Write to storage if fingerprint not yet recorded
+            try temp_await {
+                fingerprintStorage.put(package: packageIdentity,
+                                       version: version,
+                                       fingerprint: Fingerprint(origin: .sourceControl(sourceControlURL), value: revision.identifier),
+                                       observabilityScope: observabilityScope,
+                                       callbackQueue: .sharedConcurrent,
+                                       callback: $0)
+            }
+        } catch {
+            if !(error is StringError) {
```

Comment:
why is this check needed?

Reply:
To avoid logging the error thrown at L2658.

Maybe we can handle this better if the code is in `SourceControlPackageContainer` and there is a dedicated error enum, like we do with registry.

Reply:
cleaned this up in https://github.com/apple/swift-package-manager/pull/3890/commits/454a52ac8e71cc0e708ca5aa878762c16f44342d